server/src/main/java/de/uftos/repositories/solver/timefold/constraints/ConstraintDefinitionFactory.java

```diff
@@ -0,0 +1,1073 @@
+            case STUDENT_GROUP ->
+                resultSet.addAll(((StudentGroupTimefoldInstance) resource).getProvidedTagsList());
+            case ROOM -> resultSet.addAll(((RoomTimefoldInstance) resource).getProvidedTagsList());
+            case SUBJECT ->
+                resultSet.addAll(((SubjectTimefoldInstance) resource).getProvidedTagsList());
+            case GRADE ->
+                resultSet.addAll(((GradeTimefoldInstance) resource).getProvidedTagsList());
+            case TIMESLOT ->
+                resultSet.addAll(((TimeslotTimefoldInstance) resource).getProvidedTagsList());
+            default -> throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      default -> throw new IllegalStateException();
+    };
+  }
+
+  private static Function<List<ResourceTimefoldInstance>, ResourceTimefoldInstance> convertElement(
+      AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    if (ast.getToken() == UcdlToken.NUMBER || ast.getToken() == UcdlToken.SIZE) {
+      Function<List<ResourceTimefoldInstance>, Number> function = convertNumberElement(ast, params);
+      return (parameters) -> (ResourceTimefoldInstance) function.apply(parameters);
+    }
+
+    if (ast.getToken() != UcdlToken.ELEMENT) {
+      throw new IllegalStateException();
+    }
+    ElementDto element = (ElementDto) ast;
+
+    Function<List<ResourceTimefoldInstance>, ResourceTimefoldInstance> name =
+        convertValueReference(element.name(), params);
+
+    List<Function<
+        List<ResourceTimefoldInstance>,
+        Function<Set<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+        >> attributes = new ArrayList<>();
+
+    for (AbstractSyntaxTreeDto attribute : element.attributes()) {
+      attributes.add(convertAttribute(attribute, params));
+    }
+
+    return (parameters) -> {
+      Set<ResourceTimefoldInstance> elementReference = new HashSet<>();
+      elementReference.add(name.apply(parameters));
+
+      for (Function<
+            List<ResourceTimefoldInstance>,
+            Function<Set<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+            > function : attributes) {
+        elementReference = function.apply(parameters).apply(elementReference);
+      }
+      if (elementReference.size() != 1) {
+        throw new IllegalStateException();
+      }
+      return elementReference.iterator().next();
+    };
+  }
+
+  private static Function<List<ResourceTimefoldInstance>, Number> convertNumberElement(
+      AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    return switch (ast.getToken()) {
+      case NUMBER -> convertNumber(ast, params);
+      case SIZE -> convertSize(ast, params);
+      default -> throw new IllegalStateException();
+    };
+  }
+
+  private static Function<List<ResourceTimefoldInstance>, Number> convertNumber(
```

Comment:
I do quite like these `convertX` methods. They are very well contained and much better readable than those huuuuuge single methods.

Reply:
Thanks